// accord/lib/adapters/minify-html.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, MinifyHTML, W, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  W = require('when');

  _ = require('lodash');

  MinifyHTML = (function(_super) {
    var compile;

    __extends(MinifyHTML, _super);

    function MinifyHTML(compiler) {
      this.compiler = compiler;
      this.name = 'minify-html';
      this.extensions = ['html'];
      this.output = 'html';
    }

    MinifyHTML.prototype._render = function(str, options) {
      options = _.defaults(options, {
        removeComments: true,
        collapseWhitespace: true,
        removeEmptyAttributes: true
      });
      return compile((function(_this) {
        return function() {
          return _this.compiler.minify(str, options);
        };
      })(this));
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (_error) {
        err = _error;
        return W.reject(err);
      }
      return W.resolve(res);
    };

    return MinifyHTML;

  })(Adapter);

  module.exports = MinifyHTML;

}).call(this);
