// accord/lib/adapters/mustache.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Adapter, Mustache, W, fs, path, util,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  Adapter = require('../adapter_base');

  W = require('when');

  util = require('util');

  fs = require('fs');

  path = require('path');

  Mustache = (function(_super) {
    var compile;

    __extends(Mustache, _super);

    function Mustache(compiler) {
      this.compiler = compiler;
      this.name = 'mustache';
      this.extensions = ['mustache', 'hogan'];
      this.output = 'html';
    }

    Mustache.prototype._render = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.compiler.compile(str, options).render(options, options.partials);
        };
      })(this));
    };

    Mustache.prototype._compile = function(str, options) {
      return compile((function(_this) {
        return function() {
          return _this.compiler.compile(str, options);
        };
      })(this));
    };

    Mustache.prototype._compileClient = function(str, options) {
      options.asString = true;
      return this._compile(str, options).then(function(o) {
        return "new Hogan.Template(" + (o.toString()) + ");";
      });
    };

    Mustache.prototype.clientHelpers = function() {
      var runtime_path, version;
      version = require(path.join(this.compiler.__accord_path, 'package')).version;
      runtime_path = path.join(this.compiler.__accord_path, "web/builds/" + version + "/hogan-" + version + ".min.js");
      return fs.readFileSync(runtime_path, 'utf8');
    };

    compile = function(fn) {
      var err, res;
      try {
        res = fn();
      } catch (_error) {
        err = _error;
        return W.reject(err);
      }
      return W.resolve(res);
    };

    return Mustache;

  })(Adapter);

  module.exports = Mustache;

}).call(this);
